core/src/test/java/org/elasticsearch/cluster/service/ClusterServiceUtils.java

```diff
@@ -0,0 +1,101 @@
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.DummyTransportAddress;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.concurrent.CountDownLatch;
+
+import static junit.framework.TestCase.fail;
+
+public class ClusterServiceUtils {
+
+    public static ClusterService createClusterService(ThreadPool threadPool) {
+        ClusterService clusterService = new ClusterService(Settings.EMPTY, null,
+                new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS),
+                threadPool, new ClusterName("ClusterServiceTests"));
+        clusterService.setLocalNode(new DiscoveryNode("node", DummyTransportAddress.INSTANCE, Version.CURRENT));
+        clusterService.setNodeConnectionsService(new NodeConnectionsService(Settings.EMPTY, null, null) {
+            @Override
+            public void connectToAddedNodes(ClusterChangedEvent event) {
+                // skip
+            }
+
+            @Override
+            public void disconnectFromRemovedNodes(ClusterChangedEvent event) {
+                // skip
+            }
+        });
+        clusterService.setClusterStatePublisher((event, ackListener) -> {
+        });
+        clusterService.start();
+        return clusterService;
+    }
+
+    public static ClusterService createClusterService(ClusterState initialState, ThreadPool threadPool) {
+        ClusterService clusterService = createClusterService(threadPool);
+        setState(clusterService, initialState);
+        return clusterService;
+    }
+
+    public static void setState(ClusterService clusterService, ClusterState.Builder clusterStateBuilder) {
+        setState(clusterService, clusterStateBuilder.build());
+    }
+
+    public static void setState(ClusterService clusterService, ClusterState clusterState) {
+        CountDownLatch latch = new CountDownLatch(1);
+        clusterService.submitStateUpdateTask("test setting state", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                // make sure we increment versions as listener may depend on it for change
+                return ClusterState.builder(clusterState).version(currentState.version() + 1).build();
+            }
+
+            @Override
+            public boolean runOnlyOnMaster() {
+                return false;
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                latch.countDown();
+            }
+
+            @Override
+            public void onFailure(String source, Throwable t) {
+                fail("unexpected exception" + t);
+            }
+        });
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            // me
```

Comment:
I think we should either just throw `InterruptedException` or restore the interrupt flag with `Thread.currentThread.interrupt()` here (it's related to the discussion we had in #14799). Otherwise this can result in weird errors when the test infrastructure terminates hanging tests.

We could also apply a fairly conservative timeout value to `#await()` (like 1 minute) and fail the test if the return value of `#await()` is false (see [Javadoc](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CountDownLatch.html#await-long-java.util.concurrent.TimeUnit-))